In [2]:
import pandas as pd
from sklearn.svm import  SVC
import numpy as np
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn import metrics 
from sklearn import svm

In [3]:
#creating classification conditions for the ratings, so it will be divided into 'good' (ratings >=4), 'mediocre (3<ratings<4),' and 'bad'(ratings<=3).
df_goodreads_horror = pd.read_csv("HorrorSVMCookies.csv")

col = "Rating"
conditions  = [ df_goodreads_horror[col] >= 4, (df_goodreads_horror[col] < 4) & (df_goodreads_horror[col]> 3), df_goodreads_horror[col] <= 3 ]
choices     = [ "good", 'mediocre', 'bad' ]
df_goodreads_horror["Rating_class"] = np.select(conditions, choices, default=np.nan)

df_goodreads_horror

Title                Author  \
0                               The Shining (Paperback)          Stephen King   
1                                        It (Paperback)          Stephen King   
2                                   Dracula (Paperback)           Bram Stoker   
3                        Carrie (Mass Market Paperback)          Stephen King   
4                         Pet Sematary (Kindle Edition)          Stephen King   
...                                                 ...                   ...   
1243                   The Other Black Girl (Hardcover)  Zakiya Dalila Harris   
1244                  Blood Countess (Lady Slayers, #1)          Lana Popović   
1245         The Cuckoo Clock of Doom (Goosebumps, #28)            R.L. Stine   
1246                                 Färjan (Hardcover)       Mats Strandberg   
1247  Batman: Arkham Asylum - A Serious House on Ser...        Grant Morrison   

      Year  Rating   Raters Rating_class  
0     1977    4.26  1331732         good  
1     1986    4.25   979436         good  
2     1897    4.01  1119957         good  
3     1974    3.98   630669     mediocre  
4     1983    4.04   516341         good  
...    ...     ...      ...          ...  
1243  2021    3.40    59940     mediocre  
1244  2020    3.47     2136     mediocre  
1245  1995    3.64     7614     mediocre  
1246  2015    3.45     3671     mediocre  
1247  1989    4.10    67023         good  

[1248 rows x 6 columns]

In [18]:
#Splitting Dataset
feature_col = ["Rating"]
X = df_goodreads_horror[feature_col]
y = df_goodreads_horror["Rating_class"]

#Dividing the data set into train set data and test set data.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 8)

#checking the train set data
X_train
y_train

#checking the test set data
X_test
y_test


506     mediocre
111         good
93          good
1164        good
61      mediocre
          ...   
193     mediocre
236         good
599     mediocre
427         good
542     mediocre
Name: Rating_class, Length: 250, dtype: object

In [19]:
#SVM settings
svm = SVC(kernel='rbf', random_state=1, gamma=0.08, C=1)

#training the model
svm.fit(X_train, y_train)

#predicting the train set
y_pred=svm.predict(X_train)

#printing the train set accuracy score
print("Accuracy Score = "+ str(accuracy_score(y_train, y_pred))) #Accuracy Score = 0.966933867735471

Accuracy Score = 0.966933867735471


In [20]:
#predicting the test set
y_pred2=svm.predict(X_test)

#printing the train set accuracy score
print("Accuracy Score = " + str(accuracy_score(y_test, y_pred2))) #Accuracy Score = 0.976

Accuracy Score = 0.976


## Comparing the train set data and test set data 
From the svm tests that we ran, we can compare the accuracy score of the train set and test set. </br>
* the train set accuracy score is **0.966933867735471** or **97% (rounded off)** accurate.
* the test set accuracy score is **0.976** or **97%** accurate. 
Both the train set accuracy and the test set accuracy is considered very good and optimal as it is >90% accurate.</br>
The train set and test set accuracy is almost the same value (97% accurate for both). This means that our model is relatively accurate when evaluated by the SVM method. Therefore our model should be considered a good model. 